Drink docs (#305)
paritytech/ink-docs: versioned_docs/version-5.x/testing/drink.md

---
title: DRink!
hide_title: true
slug: /basics/contract-testing/drink
---

<img src="/img/title/testing1.svg" className="titlePic" />

# DRink!

> Apart from the core ink! testing framework, we also have the [DRink!](https://github.com/inkdevhub/drink) library.

Reviewer: ```suggestion
Apart from the core ink! testing framework, Cardinal Cryptogtaphy also provides the [DRink!](https://github.com/inkdevhub/drink) library.
```

> It provides an intermediate solution between integration and E2E testing.

Reviewer: ```suggestion
It offers an intermediate solution between integration and E2E testing.
```

> Briefly speaking, DRink! maintains full in-memory blockchain state and allows you to directly interact with it.

Reviewer: ```suggestion
Intuitively, DRink! maintains a full in-memory blockchain state and allows you to interact with it directly.
```

This gives a notably robust methodology in contrast to the off-chain engine facilitated by the `ink::test` macro.
Nevertheless, it is essential to note that the absence of the entire node layer makes the environment less realistic compared to the comprehensive setup employed in the end-to-end tests.

## Comparison to the other testing strategies

> To better assess when DRINK performs better than other testing methods, it is crucial to gain a deeper understanding of the consequences of its unique design and the trade-offs it entails.

Reviewer: ```suggestion
To better assess when DRink! performs better than other testing methods, it is crucial to gain a deeper understanding of the consequences of its unique design and the trade-offs it entails.
```

Author: :white_check_mark:

1. **Speed:** since DRink! doesn't spawn any node or background process, everything happens locally, within the testing thread.
This means that the execution can be synchronous and significantly faster than the E2E tests.
Therefore, its performance should be comparable to the unit or integration testing.
Also, there is no block production or finalization delay, which is a noticeable factor in the E2E tests.
Thanks to that, we can launch long-running simulations in a reasonable time.

2. **Testing multiple contracts:** since we are working with a full blockchain state, we can perform any interaction with the contracts, which includes working with mutliple contracts at the same time.

> This is of course the same as in the E2E tests, but not possible in neither unit nor integration tests.

Reviewer: ```suggestion
Of course, this is the same as in the E2E tests, but it is not possible in either the unit or integration tests.
```

Author: :white_check_mark:

> 3. **Working with arbitrary runtimes:** similarly to the E2E tests, where we can spawn any node with customized runtime (Polkadot's term for the state transition function), in DRink! tests we can work with any blockchain we want.

Reviewer: ```suggestion
3. **Working with arbitrary runtimes:** similarly to the E2E tests, where we can spawn any node with customized runtime (Polkadot's term for the state transition function), in DRink! tests we can work with any blockchain runtime we want.
```

Author: :white_check_mark:

> 4. **Full control over runtime state:** we hold the state of the blockchain in hand, so we can easily manipulate it in any way we want.

Reviewer: ```suggestion
4. **Full control over runtime state:** we hold the state of the blockchain and exercise full control over it, so we can easily manipulate it however we want.
```

   - **enhanced debugging and call tracing:** you can get more insights into the contract execution process, like stack trace, debug buffers and more.

Nevertheless, there are some drawbacks of DRink! as well:

1. **No node layer:** since we don't spawn any node, we don't have access to the node layer, which means that we can't test any node-related functionality, like RPC calls, block production, etc.

2. **Artificial, isolated environment:** this is the main trade-off of DRink!.
It might give a false sense of security, while in the real environment, the contract could behave differently.
The discrepancy can be mitigated by a careful and precise simulation and setup of the environment, but it is still a factor to consider.

3. **No typed contract API:** currently, DRink! works with string-encoded arguments and values, which means that we lose the type safety and convenience that was present in the other testing frameworks.
Fortunately, this is going to change soon, as there is an ongoing effort to integrate it with [ink-wrapper](https://github.com/Cardinal-Cryptography/ink-wrapper) library.

> ## When to use DRink!?

Reviewer: ```suggestion
## When to use `DRink!`?
```

Author: :white_check_mark:

    #[drink::test]
    fn deploy_and_call_a_contract(mut session: Session) -> Result<(), Box<dyn Error>> {
        let result: bool = sesison
            .deploy_bundle_and(BundleProvider::local(), "new", &["true"], NO_SALT, NO_ENDOWMENT)?
            .call_and("flip", NO_ARGS, NO_ENDOWMENT)?
            .call_and("flip", NO_ARGS, NO_ENDOWMENT)?
            .call_and("flip", NO_ARGS, NO_ENDOWMENT)?
            .call("get", NO_ARGS, NO_ENDOWMENT)??;
        assert_eq!(result, false);
    }
}
```

You can check some helpful and verbose examples [here](https://github.com/inkdevhub/drink/tree/main/examples), including the [**quick start guide**](https://github.com/inkdevhub/drink/tree/main/examples/quick-start-with-drink).

### As an alternative backend to ink!'s E2E testing framework

DRink! is already integrated with the ink! framework and can be used as a drop-in replacement for the standard E2E testing environment.
Just use corresponding argument in the test macro:
```rust
#[ink_e2e::test(backend(runtime_only()))]
```
to your test function and you have just switched from E2E testcase to DRink!-based one, that doesn't use any running node in the background!

> For a full example check out [ink! repository](https://github.com/paritytech/ink/blob/master/integration-tests/e2e-runtime-only-backend/lib.rs).

Reviewer 2: I'll update this link to point to https://github.com/paritytech/ink-examples once we've updated that repo to 5.0. It'll then contain the content of the ink! repositories `integration-tests` folder.
